frontends/pytorch/csrc/c10_dispatch/acap_dispatch.cpp

```diff
@@ -62,23 +106,170 @@ std::shared_ptr<AcapController> AcapController::getCurrent() {
-    // Capture the dispatch.
-    std::stringstream sout;
-    sout << "CAPTURE: " << opHandle.schema() << "\n";
-    current->captureLog.push_back(sout.str());
+  if (!current) {
+    current->redispatch(opHandle, stack);
+    return;
   }
+  current->fallbackKernelImpl(opHandle, stack);
+}
 
+void AcapController::redispatch(const c10::OperatorHandle &opHandle,
+                                c10::Stack *stack) {
+  // Exclude recursive dispatch to this kernel.
+  c10::impl::ExcludeDispatchKeyGuard exclusion(kAcapDispatchKey);
+  // Passthrough.
   auto &dispatcher = c10::Dispatcher::singleton();
   dispatcher.callBoxed(opHandle, stack);
 }
 
+void AcapController::fallbackKernelImpl(const OperatorHandle &opHandle,
+                                        Stack *stack) {
+  verifyHasNotReturned();
+  // Exclude recursive dispatch to this kernel.
+  c10::impl::ExcludeDispatchKeyGuard exclusion(kAcapDispatchKey);
+
+  const FunctionSchema &schema = opHandle.schema();
+
+  // Check for unsupported.
+  if (schema.is_vararg() || schema.is_varret()) {
+    throw std::invalid_argument(
+        "Cannot capture ops with variable arguments or returns");
+  }
+
+  // TODO: Extract actual location from stack.
+  MlirContext context = funcBuilder->getContext();
+  MlirLocation loc = mlirLocationUnknownGet(context);
+  OperationStateHolder stateHolder("torch.kernel_call", loc);
+
+  // Add the kernel_name attribute.
+  auto kernelName = schema.name();
+  MlirNamedAttribute kernelNameAttr = mlirNamedAttributeGet(
+      "kernel_name",
+      mlirStringAttrGet(context, kernelName.size(), kernelName.data()));
+  mlirOperationStateAddAttributes(&stateHolder.state, 1, &kernelNameAttr);
+
+  // Map arguments to operands.
+  // This must be accumulated into the OperationState prior to re-dispatch
+  // since the stack is modified at that point.
+  size_t argCount = schema.arguments().size();
+  assert(stack->size() >= argCount && "stack too short");
+  llvm::SmallVector<MlirValue, 4> operands;
+  for (auto argIt = stack->end() - argCount; argIt != stack->end(); ++argIt) {
+    MlirValue mlirValue = mapIValueToMlirValue(loc, *argIt);
+    // TODO: Add upstream mlirValueIsNull
+    if (!mlirValue.ptr) {
+      std::stringstream out;
+      out << "Unsupported capture value passed to kernel (" << argIt->tagKind()
+          << "): " << *argIt;
+      throw std::invalid_argument(out.str());
+    }
+    operands.push_back(mlirValue);
+  }
+  mlirOperationStateAddOperands(&stateHolder.state, operands.size(),
+                                operands.data());
+
+  // Invoke the original kernel.
+  redispatch(opHandle, stack);
+
+  // Map returns to results.
+  size_t returnCount = schema.returns().size();
+  assert(stack->size() >= returnCount && "stack too short");
+  llvm::SmallVector<MlirType, 4> resultTypes;
+  llvm::SmallVector<std::pair<size_t, at::Tensor>, 4> resultIndexToTensorMap;
+  for (auto returnIt = stack->end() - returnCount; returnIt != stack->end();
```

Comment:
Is the first argument top of stack or not? This code suggests it is not, and the torch.kernel_call docs need to be udpated.

Reply:
Seems not. I had it reversed and that turned out to not be right.